### PR TITLE
feat: Array comprehension

### LINKED ISSUE #612: Array comprehension
Blocked by #610

### PR DESCRIPTION
**Pull request by mark-koch**
Adds array comprehensions of the form `array(i for i in range(10))`.

This is the base PR that adds new AST nodes for array comprehensions and general generator expressions. Checking and lowering logic follows in subsequent PRs targetting this branch:
* #614
* #615 
* #616

Closes #612

### CODE REVIEW

File: guppylang/definition/custom.py [hunk 1]
```diff
@@ -261,12 +261,16 @@ def _setup(self, ctx: Context, node: AstNode, func: CustomFunctionDef) -> None:
...
     def check(self, args: list[ast.expr], ty: Type) -> tuple[ast.expr, Subst]:
         """Checks the return value against a given type.
 
         Returns a (possibly) transformed and annotated AST node for the call.
         """
+        from guppylang.checker.expr_checker import check_type_against
+
+        expr, res_ty = self.synthesize(args)
+        subst, _ = check_type_against(res_ty, ty, self.node)
+        return expr, subst
```
Comment (ss2165) [inline guppylang/definition/custom.py:273]: not clear to me why this change is related here

File: guppylang/prelude/builtins.py [hunk 2]
```diff
@@ -559,6 +561,27 @@ def __len__(self: array[L, n]) -> int: ...
     def __new__(): ...
 
 
+@guppy.extend_type(sized_iter_type_def)
+class SizedIter:
+    """A wrapper around an iterator type `T` promising that the iterator will yield
```
Comment (ss2165) [inline guppylang/prelude/builtins.py:566]:
```suggestion
    """A wrapper around an iterator type `L` promising that the iterator will yield
```
?

File: guppylang/prelude/builtins.py [hunk 3]
```diff
@@ -559,6 +561,27 @@ def __len__(self: array[L, n]) -> int: ...
...
+    def __new__(iterator: L @ owned) -> "SizedIter[L, n]":  # type: ignore[type-arg]
+        """Casts an iterator into a `SizedIter`."""
+
+    @guppy.custom(NoopCompiler())
+    def unwrap_iter(self: "SizedIter[L, n]" @ owned) -> L:
+        """Extracts the actual iterator."""
+
+    @guppy.custom(NoopCompiler())
+    def __iter__(self: "SizedIter[L, n]" @ owned) -> L:
+        """Extracts the actual iterator."""
```
Comment (ss2165) [inline guppylang/prelude/builtins.py:582]: what's the difference between these? (is there one?)

**Review by ss2165 — COMMENTED** [review]
👍

File: guppylang/definition/custom.py (hunk 1 above)
Comment (mark-koch) [reply guppylang/definition/custom.py]: This was added in #610. Github showed the diff because of a merge conflict, sorry 😅

File: guppylang/prelude/builtins.py (hunk 3 above)
Comment (mark-koch) [reply guppylang/prelude/builtins.py]: It's the same, I thought we might want to have a non-dunder method to get the wrapped iterator?

File: guppylang/compiler/expr_compiler.py [hunk 4]
```diff
@@ -492,11 +492,8 @@ def visit_DesugaredArrayComp(self, node: DesugaredArrayComp) -> Wire:
         assert isinstance(array_ty, OpaqueType)
         array_var = Variable(next(tmp_vars), array_ty, node)
         count_var = Variable(next(tmp_vars), int_type(), node)
-        hugr_elt_ty = (
-            ht.Option(node.elt_ty.to_hugr())
-            if node.elt_ty.linear
-            else node.elt_ty.to_hugr()
-        )
+        # See https://github.com/CQCL/guppylang/issues/629
```
Comment (ss2165) [inline guppylang/compiler/expr_compiler.py:493]: not sure about the relevance of this link?

**Review by ss2165 — APPROVED** [review]

File: guppylang/compiler/expr_compiler.py (hunk 4 above)
Comment (mark-koch) [reply guppylang/compiler/expr_compiler.py:493]: This is to explain why we use `Option` for both linear and classical arrays
